runtime/service.go

```diff
@@ -48,20 +57,72 @@ import (
 	"github.com/firecracker-microvm/firecracker-containerd/eventbridge"
 	"github.com/firecracker-microvm/firecracker-containerd/internal"
 	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	"github.com/firecracker-microvm/firecracker-containerd/runtime/firecrackeroci"
 )
 
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
 const (
 	defaultVsockPort     = 10789
 	supportedMountFSType = "ext4"
+
+	vmIDEnvVarKey = "FIRECRACKER_VM_ID"
+
+	varRunDir                  = "/var/run/firecracker-containerd"
+	firecrackerSockName        = "firecracker.sock"
+	firecrackerLogFifoName     = "fc-logs.fifo"
+	firecrackerMetricsFifoName = "fc-metrics.fifo"
+
+	// TODO these strings are hardcoded throughout the containerd codebase, it may
+	// be worth sending them a PR to define them as constants accessible to shim
+	// implementations like our own
+	shimAddrFileName = "address"
+	shimLogFifoName  = "log"
+	ociConfigName    = "config.json"
 )
 
 // implements shimapi
 type service struct {
 	server        *ttrpc.Server
 	id            string
+	vmID          string
 	publish       events.Publisher
 	eventExchange *exchange.Exchange
 
+	// vmIDDir holds files, sockets and FIFOs scoped to a single given VM
+	vmIDDir string
+
+	// bundleDir is created by containerd and contains configuration scoped to a
+	// single given container
+	bundleDir string
+
+	// shimAddrFilePath is the path to the shim address file as found in the vmIDDir
+	shimAddrFilePath string
+
+	// bundleAddrFilePath is the path to the address file as found in the bundleDir.
+	// Even though the shim address is set per-VM, not per-container, containerd expects
+	// to find the shim addr file in the bundle dir, so we still have to create it or
+	// symlink it to the shimAddrFilePath
+	bundleAddrFilePath string
+
+	// shimLogFifoPath is a path to a FIFO for writing shim logs as found in the vmIDDir.
+	shimLogFifoPath string
+
+	// bundleLogFifoPath is a path to a FIFO for writing shim logs as found in the bundleDir.
+	// It is the path created by containerd for us, the shimLogFifoPath is just a symlink to one.
+	bundleLogFifoPath string
```

Comment:
It looks like `bundleLogFifoPath` and `shimLogFifoPath` are only required for start up logic (specifically creating symlinks). Can we evaluate those just once inside `createSymlinkXXX` and don't keep them on instance level?

Reply:
We can, the reason I wanted them on the instance level is that it centralizes our definition of paths in one place (which helps a lot for clarity I think) and in case they ever need to be used in more than one place in the code, they will already be accessible.

Given the cost of storing 2 strings is minimal and they are expected to be constant for the lifetime of the service, I'd prefer to keep them in the instance level unless there's other considerations I'm not thinking of, let me know.

Reply:
It's more about readability/maintainability. I would rather avoid polluting struct definitions with fields that are not going to be used. In case when we need to access in more than one place, I would rather use `getXxxPath` instead (but in this case its only one usage).

Reply:
How do you feel about updating to use `get*Path` methods for all the paths? I didn't do this originally because it felt silly to have a function for returning a value that's a known constant after initialization, but in retrospect it does make those values more immutable, which is nice.

I still don't want to just inline any paths inside the function where they are used because it makes it hard for someone unfamiliar with the code to know what the overall directory structure is and all the paths the code interacts with. They have to search all over the place rather than look in one central place where all paths are defined. This is something that's frustrated me on multiple occasions when reading various other codebases.

Reply:
yep, `get*Path` is fine.